### PR TITLE
New auth page

### PR DESCRIPTION
**Pull request by pjanik**
This PR adds a new page with login form (`/auth/login`) which is used during SSO login (e.g. when user comes from LARA and he's not logged in Portal yet).

I've recreated Ethan's draft: http://staff.concord.org/emcelroy/login/index.html
Ethan's used a bit different technique to create placeholders, but it didn't work too well for me with autocomplete in Chrome. I've just used `placeholder()` polyfill that we use in other places in Portal.

To avoid redirecting to the home page when user provides wrong username or password, I've had to change a bit `lib/custom_failure#redirect_url`. Now, when login fails, user gets redirected back to the original page (instead of home page).

Error message wasn't designed in the draft, so I just implemented this:
screen shot 2016-02-18 at 18 01 31
I think it's pretty trivial, so probably it's not worth asking Ethan to look at it (but I won't feel offended if someone thinks he should do it :wink:).

### CODE REVIEW

File: app/views/auth/login.haml [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+      }
+      .error-message {
+        color: #860024;
+      }
+    %body
+    #page-wrap
+      = form_tag session_path(:user), :id => "login-form", :class => @error.present? ? "with-errors" : "" do
+        = image_tag "cc-logo.png", :title => "The Concord Consortium", :alt => "The Concord Consortium"
+        %p
+          Log in with your Learn Portal account
```
Comment (scytacki) [inline app/views/auth/login.haml]: This code will get applied to other portals besides the learn.concord.org portal. So you should use the site_name from settings.yml and that should work.

File: lib/custom_failure.rb [hunk 2]
```diff
@@ -13,7 +13,11 @@ def redirect
   end
 
   def redirect_url
-    root_path
+    if request.get?
+      attempted_path
+    else
+      request.referrer
+    end
```
Comment (scytacki) [inline lib/custom_failure.rb]: Aaron was working on this before in order to deal with some SSO issues:
https://github.com/concord-consortium/rigse/pull/52
I think the main things to test are:
1. what happens when your portal session is expired and you try to log in through the portal from LARA or some other OAuth client.
2. what happens when your portal session is expired and you try to go somewhere that requires authentication.
3. what happens when your portal session is expired and you try to save a form of some kind (POST request). Also check what happens when this form is submitted via XHR.
4. recheck #2 and #3 when the cookie has been completely deleted (which should trigger a warden message other than `:timeout`)

**Comment by pjanik** [comment]
@scytacki, 

> what happens when your portal session is expired and you try to log in through the portal from LARA or some other OAuth client.

It still works. `CustomFailure#redirect` calls:
https://github.com/concord-consortium/rigse/blob/master/lib/custom_failure.rb#L9
`attempted_path` is equal to `/auth/concord_id/authorize?response_type=code&client_id=localhost&redirect_uri=http%3A%2F%2Floc(...)`
We are redirected back to the `AuthController` as an anonymous user and follow “standard” path which redirects us to the new login page (using `#verify_logged_in` method which I’ve modified in this PR).

> what happens when your portal session is expired and you try to go somewhere that requires authentication.

**Example**: session is expired, but you try to access `/admin` page.
It still works. `CustomFailure#redirect` calls:
https://github.com/concord-consortium/rigse/blob/master/lib/custom_failure.rb#L9
`attempted_path` is equal to `/admin`. We are be redirected back to `/admin` as an anonymous user. Then, Pundit takes care of the rest:
https://github.com/concord-consortium/rigse/blob/master/app/controllers/home_controller.rb#L7-L12
We are redirected to the home page again. It may vary from page to page (depending on implementation in given controller), but basically it’s Pundit responsibility to redirect us to an appropriate place. `CustomFailure#redirect_url` is not called at all in this case. This method is only called by the original implementation of redirect (`Devise::FailureApp#redirect`). We call `super` only when warden message is different from `:timeout`.

> what happens when your portal session is expired and you try to save a form of some kind (POST request). Also check what happens when this form is submitted via XHR.

**Non-XHR POST example**: session is expired, but you try to create a new class.
It still works. It’s bascially the same case as 2 (and similar to 1). Warden message is `:timeout`, so we are redirected to the original URL as an anonymous user and Pundit takes care of the rest.
 
**XHR POST example**: session is expired, but you try to submit admin project edit form.
 It still works. `CustomFailure#redirect` follows `http_auth?` code path and returns 401 error:
https://github.com/concord-consortium/rigse/blob/master/lib/custom_failure.rb#L28
`CustomFailure#redirect_url` is not called at all in this case.
It doesn’t really look good in Portal. An error message is displayed instead of the form (I think it’s not intentional and specific to this case), but the page is not refreshed and it looks like you’re still logged in. You need to refresh page. I guess it might be worse in some other cases. But this PR doesn’t seem to change anything.

> recheck #2 and #3 when the cookie has been completely deleted (which should trigger a warden message other than :timeout)

I don't think that other warden message will be used. I guess it’s just an anonymous user trying to access various resources, so it’s more Pundit failure, not warden. `CustomFailure` doesn’t seem to be used at all. I’ve also tried to delete session manually from the DB, but it seems to be the same case. But I've tested that too:

**2**: I’m redirected to home page by Pundit.

**3 XHR**: I get an empty page with “send us email” link. It looks and feels wrong, but there is no regression.

**3 Non-XHR**: I can see alert with `Anonymous can't create classes. Please log in and try again.` message. The form is still visible in the background. Feels weird, but again, not really related to warden or this PR.

I can’t really think of other warden message type. What else can happen? Any ideas?
Can't find clear list in docs or source. I’ve been searching for `throw :warden` in Devise repo and found only two cases:

`throw :warden, :scope => scope, :message => record.inactive_message`
`throw :warden, :scope => scope, :message => :timeout`

`:timeout` is already covered, and the second one probably isn’t really relevant to our User model (as it’s not “activatable”).

Uff, hope that helps. At least I understand that a bit more now.

**Comment by scytacki** [comment]
Looks good to me. Thanks for testing it all out.

**Pull request merged**